### PR TITLE
[5X only]Should not modify pointer's memory after free it in CopyFromDispatch

### PR DESCRIPTION
**Pull request by JunfengYang**
Since commit be82a3d6761b99e2fbe8aa169728812af2f81bd9, we alloc
part_distData in the per tuple memory context. But we still try
to modify it's values after reset the memory context.
This will lead to memory corruption.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Comment by asimrp** [comment]
The change looks subtle!  Someone refactoring this code in future may easily revert this patch and bring back the use-after-free bug.  How did you find it?  Is it possible to add a test to catch it?

Taking a closer look, `GetTargetSeg()` is the only user of `part_distData`.  Would it be better to allocate it inside this function?

**Comment by JunfengYang** [comment]
@asimrp `GetTargetSeg` is also called in `CopyFrom` which actually called under COPY ON SEGMENTS. And the distribution policy check is different from `CopyFromDispatch`. To move the `distData` initiation into GetTargetSeg` is not a good solution which makes logic complex and confuses.
And right now, the current change moves the `part_distData` allocation just before calling `GetTargetSeg`.
I don't think in the future we will modify `part_distData` before the below code which does the allocation.
```
                                        if (estate->es_result_partitions)
					{
						getAttrContext->tupDesc = tupDesc;
						getAttrContext->attr = attr;
						getAttrContext->num_phys_attrs = num_phys_attrs;
						getAttrContext->attr_offsets = attr_offsets;
						getAttrContext->nulls = nulls;
						getAttrContext->values = values;
						getAttrContext->cdbCopy = cdbCopy;
						getAttrContext->original_lineno_for_qe =
							original_lineno_for_qe;
						part_distData = GetDistributionPolicyForPartition(
							        cstate, estate, partitionData,
							        distData->hashmap, distData->p_attr_types,
							        getAttrContext, oldcontext);
					}
					else
						part_distData = (GpDistributionData *) palloc0(sizeof(GpDistributionData));
```

For the test part, as long as the freed memory not allocated to others, there will be no error. So it's hard to write a test to cover it.
And even it does allocate to others, we may still have no errors.

Find the issue is really hard, I try to modify the code with logs to expose the bug, but it didn't expose it every time.
But sometimes the error memory address is the same, so I just use gdb to monitor the address's life cycle several times and then find the root cause.

**Comment by JunfengYang** [comment]
Move the `GpDistributionData *part_distData` declaration in the while loop may resolve your concern.

File: src/backend/commands/copy.c [hunk 1]
```diff
@@ -4005,6 +4004,8 @@ CopyFromDispatch(CopyState cstate)
 							        distData->hashmap, distData->p_attr_types,
 							        getAttrContext, oldcontext);
 					}
+					else
+						part_distData = (GpDistributionData *) palloc0(sizeof(GpDistributionData));
```
Comment (asimrp) [inline src/backend/commands/copy.c:4008]: A comment to explicitly state that this allocation needs to happen in per-tuple memory context will be helpful.

**Review by asimrp — APPROVED** [review]
Thank you @JunfengYang for the explanation.

**Pull request merged**